### ISSUE TITLE
External cray-mpich provides incorrect spec['mpi'].prefix.include path

### ISSUE THREAD

**Issue by wspear** [LLNL/spack#25612]
### Steps to reproduce

This is observed on the spock system, a cray early-access machine at ORNL. It causes several mpi dependent package builds to fail including TAU, which I will use as an example. 

packages.yaml looks like:

```
packages:
    mpich:
      buildable: false
      externals:
      - spec: mpich@8.1.7
        modules:
        - cray-mpich/8.1.7
```

The TAU package sets up its mpi configuration like this:
```
            options.append("-mpiinc=%s" % spec['mpi'].prefix.include)
            options.append("-mpilib=%s" % spec['mpi'].prefix.lib)
```

Which translates to this in the actual configure invocation:
`'-mpiinc=/opt/cray/pe/mpich/8.1.7/gtl/include' '-mpilib=/opt/cray/pe/mpich/8.1.7/gtl/lib'`

However, /opt/cray/pe/mpich/8.1.7/gtl/include does not exist. The include files for this mpi install are located in compiler specific subdiriectories of /opt/cray/pe/mpich/8.1.7/ofi and /opt/cray/pe/mpich/8.1.7/ucx. (e.g. /opt/cray/pe/mpich/8.1.7/ofi/gnu/9.1/include)

This makes the system-provided mpich impossible to use as an external, at least for packages that use spec['mpi'].prefix.include to get the include path.

### Information on your system

* **Spack:** 0.16.2-4019-80713e234c
* **Python:** 3.6.13
* **Platform:** cray-sles15-zen2
* **Concretizer:** original

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by christoph-conrads** [comment]
Hi, I was struggling with MPI on another machine and this is detailed in issue #12520 at the very bottom (this issue is only superficially about HDF5) and ended up writing documentation on how I _think_ Spack uses modules in pull request #25666 (attention: not reviewed yet). Could please show me the output of `module show cray-mpich/8.1.7`?

**Comment by wspear** [comment]
After a lot of documentation and license info the setup portion of 'module show' gives:

```
whatis("cray-mpich - Cray MPICH Message Passing Interface")
setenv("CRAY_MPICH_VER","8.1.7")
setenv("CRAY_MPICH_VERSION","8.1.7")
setenv("CRAY_MPICH_ROOTDIR","/opt/cray/pe/mpich/8.1.7")
setenv("CRAY_MPICH_BASEDIR","/opt/cray/pe/mpich/8.1.7/ofi")
setenv("PE_MPICH_MODULE_NAME","cray-mpich")
setenv("CRAY_MPICH_DIR","/opt/cray/pe/mpich/8.1.7/ofi/crayclang/10.0")
setenv("CRAY_MPICH_PREFIX","/opt/cray/pe/mpich/8.1.7/ofi/crayclang/10.0")
setenv("MPICH_DIR","/opt/cray/pe/mpich/8.1.7/ofi/crayclang/10.0")
setenv("PE_MPICH_PKGCONFIG_VARIABLES","PE_MPICH_GTL_DIR_@accelerator@:PE_MPICH_GTL_LIBS_@accelerator@")
setenv("PE_MPICH_GTL_DIR_amd_gfx906","-L/opt/cray/pe/mpich/8.1.7/gtl/lib")
setenv("PE_MPICH_GTL_DIR_amd_gfx908","-L/opt/cray/pe/mpich/8.1.7/gtl/lib")
setenv("PE_MPICH_GTL_DIR_nvidia70","-L/opt/cray/pe/mpich/8.1.7/gtl/lib")
setenv("PE_MPICH_GTL_DIR_nvidia80","-L/opt/cray/pe/mpich/8.1.7/gtl/lib")
setenv("PE_MPICH_GTL_LIBS_amd_gfx906","-lmpi_gtl_hsa")
setenv("PE_MPICH_GTL_LIBS_amd_gfx908","-lmpi_gtl_hsa")
setenv("PE_MPICH_GTL_LIBS_nvidia70","-lmpi_gtl_cuda")
setenv("PE_MPICH_GTL_LIBS_nvidia80","-lmpi_gtl_cuda")
prepend_path("PE_MPICH_GENCOMPILERS_CRAYCLANG","10.0")
prepend_path("PE_MPICH_FIXED_PRGENV","CRAYCLANG")
prepend_path("PE_CRAYCLANG_FIXED_PKGCONFIG_PATH","/opt/cray/pe/mpich/8.1.7/ofi/crayclang/10.0/lib/pkgconfig")
prepend_path("PE_PKGCONFIG_PRODUCTS","PE_MPICH")
prepend_path("PE_MPICH_PKGCONFIG_LIBS","mpich")
prepend_path("PE_MPICH_FORTRAN_PKGCONFIG_LIBS","mpichf90")
prepend_path("PE_PKGCONFIG_LIBS","mpich")
prepend_path("PE_FORTRAN_PKGCONFIG_LIBS","mpichf90")
setenv("PE_PERFTOOLS_MPICH_LIBDIR","/opt/cray/pe/mpich/8.1.7/ofi/crayclang/10.0/lib")
prepend_path("MANPATH","/opt/cray/pe/mpich/8.1.7/man/mpich")
prepend_path("MANPATH","/opt/cray/pe/mpich/8.1.7/ofi/man")
prepend_path("PATH","/opt/cray/pe/mpich/8.1.7/ofi/crayclang/10.0/bin")
prepend_path("CRAY_LD_LIBRARY_PATH","/opt/cray/pe/mpich/8.1.7/ofi/crayclang/10.0/lib")
setenv("CRAY_LMOD_MPI","cray-mpich/8.0")
prepend_path("MODULEPATH","/opt/cray/pe/lmod/modulefiles/mpi/crayclang/10.0/ofi/1.0/cray-mpich/8.0")
prepend_path("MODULEPATH","/opt/cray/pe/lmod/modulefiles/cncm/crayclang/10.0/ofi/1.0/x86-rome/1.0/cray-mpich/8.0"
```

**Comment by christoph-conrads** [comment]
Could please try to add the following line to `~/.spack/packages.yaml` and attempt to rebuild?
```diff
diff --git a/.spack/packages.yaml b/.spack/packages.yaml
index 1c9bf96..bf2c54a 100644
--- a/.spack/packages.yaml
+++ b/.spack/packages.yaml
@@ -3,5 +3,6 @@ packages:
     buildable: false
     externals:
     - spec: mpich@8.1.7
+      prefix: /opt/cray/pe/mpich/8.1.7
       modules:
       - cray-mpich/8.1.7
```

On CentOS 7 I observed that Spack was computing the wrong prefix in the presence of nonstandard paths, i.e., instead of having a directory with the folders `bin`, `lib`, `include`, and so on, the OpenMPI package on CentOS 7 is found in `/usr/lib64/openmpi/bin`, `/usr/lib64/openmpi/lib`, and `/usr/include/openmpi-x86_64`. The fix was to set `/usr` as a prefix. Following the same logic, my guess is to set the prefix above.

**edit**

If this does not work, I would try to use the value of `CRAY_MPICH_PREFIX`: `/opt/cray/pe/mpich/8.1.7/ofi/crayclang/10.0`.

**Comment by wspear** [comment]
Setting the packages.yaml prefix to CRAY_MPICH_PREFIX seems to have it working. Thanks! Do you know if there's any likelihood of automating this on the spack side?

**Comment by christoph-conrads** [comment]
> Setting the packages.yaml prefix to CRAY_MPICH_PREFIX seems to have it working. Thanks! Do you know if there's any likelihood of automating this on the spack side?

For my PR #25666 I looked at the code in [`lib/spack/spack/utils/module_cmd.py`](https://github.com/spack/spack/blob/f5ab3ad82a62585589296ff37ebf7af1aed8303f/lib/spack/spack/util/module_cmd.py#L140-L150) and the responsible function is based on heuristics counting the number of occurrences of paths. This could be automated (with module name `cray-mpich`, the prefix in `CRAY_MPICH_PREFIX` is not hard to derive) but whoever wrote the code had his reasons for the educated guesses in it. I suggest to talk to one of the Spack developers, to open a new issue, or to leave a comment in #25666.

**Comment by jgalarowicz** [comment]
I also have been having a problem building on the Cray.   After seeing this I added a modules setting (I had the prefix) to the externals file and now I can build our MPI collectors on a Cray.   The paths to the MPI installation were missing.

**Comment by kwryankrattiger** [comment]
This really should be automated, the `CRAY_MPICH_PREFIX` will change depending on the module configuration, so putting it into the package.yaml is not going to be reliable.

I am also on spock, another work around that will do what appears to be the right thing is to call the `cray-mpich@8.1.7` `mpich@3.4.2` in the package spec. Thanks @chuckatkins for this!

```yaml
package:
  mpich:
    externals:
    - spec: mpich@3.4.2
      modules:
      - cray-mpich/8.1.7
      - craype-network-ofi
```

**Comment by lukebroskop** [comment]
@wspear, I think you want your packages.yaml to use the **cray-mpich** package (not **mpich**).

e.g.
```
packages:
  cray-mpich:
    buildable: false
    externals:
    - spec: cray-mpich@8.1.7
      modules:
      - cray-mpich/8.1.7
```

**Comment by haampie** [comment]
@christoph-conrads yes, that heuristic is an ugly hack and doesn't make sense. Unfortunately there are no rules to module files, so it's hard to extract the prefix path from them.

But do note that the `cray-mpich` package has an `external_prefix` property where the actual prefix is pulled from:

```
    @property
    def external_prefix(self):
        mpich_module = module("show", self.modname).splitlines()

        for line in mpich_module:
            if "CRAY_MPICH_DIR" in line:
                return get_path_args_from_module_line(line)[0]

        # Fixes an issue on Archer2 cray-mpich/8.0.16 where there is
        # no CRAY_MPICH_DIR variable in the module file.
        for line in mpich_module:
            if "CRAY_LD_LIBRARY_PATH" in line:
                libdir = get_path_args_from_module_line(line)[0]
                return os.path.dirname(os.path.normpath(libdir))
```

and as you can see, it's again terribly system dependent, since module files aren't uniform across cray machines.

You should be using modules not prefixes, because in some cases the module sets LD_LIBRARY_PATH too, because cray doesn't always use rpaths... :upside_down_face:

**Comment by kwryankrattiger** [comment]
I had tried "correct" approach by adding `cray-mpich` as an external, buildable=false, module package but the `cray-mpich` package returns a `None` for the `prefix`. And then it dies on trying to index `None` with an out of range on spack. There is a `CRAY_MPICH_DIR` defined in the module file, so maybe there is a bug in the `get_path_args_from_module_line` routine?

**Comment by kwryankrattiger** [comment]
> @wspear, I think you want your packages.yaml to use the **cray-mpich** package (not **mpich**).
> 
> e.g.
> 
> ```
> packages:
>   cray-mpich:
>     buildable: false
>     externals:
>     - spec: cray-mpich@8.1.7
>       modules:
>       - cray-mpich/8.1.7
> ```

This is what I initially tried and it failed. The `mpich` hack is a work around that problem.

**Comment by wspear** [comment]
Providing mpi with cray-mpich is preventing my environment from concretizing, presumably because one of the top level specs is forcing mpich. I haven't been able to track down which one yet.

```
==> Error: The spec
    'mpich ^findutils ^pkgconf'
    is configured as not buildable, and no matching external installs were found
```

**Comment by lukebroskop** [comment]
@wspear , what spec are you trying to concretize? 
I just tried Tau and it concretized (this is from Spock)

```
==> Concretized tau%cce+mpi ^elfutils%gcc
 -   bvhc7hc  tau@2.30.1%cce@12.0.1~adios2+binutils~comm~craycnl~cuda+elf+fortran~gasnet+io~level_zero+libdwarf+libunwind~likwid+mpi~ompt~opari~opencl~openmp+otf2+papi+pdt~phase~ppc64le~profileparam+pthreads~python~rocm~rocprofiler~scorep~shmem~sqlite~x86_64 arch=cray-sles15-zen2
[+]  vi6s6bw      ^binutils@2.33.1%cce@12.0.1~gas~gold+headers~interwork~ld+libiberty~lto+nls+plugins libs=shared,static arch=cray-sles15-zen2
[+]  lj6676v          ^diffutils@3.7%cce@12.0.1 arch=cray-sles15-zen2
[+]  inpdowd              ^libiconv@1.16%cce@12.0.1 libs=shared,static arch=cray-sles15-zen2
[+]  a4hlix4          ^gettext@0.21%cce@12.0.1+bzip2+curses+git~libunistring+libxml2+tar+xz arch=cray-sles15-zen2
[+]  3fdmu2n              ^bzip2@1.0.8%cce@12.0.1~debug~pic+shared arch=cray-sles15-zen2
[+]  a5phfbk              ^libxml2@2.9.12%cce@12.0.1~python arch=cray-sles15-zen2
[+]  npj2cpm                  ^pkgconf@1.8.0%cce@12.0.1 arch=cray-sles15-zen2
[+]  ypnjcgn                  ^xz@5.2.5%cce@12.0.1~pic libs=shared,static arch=cray-sles15-zen2
[+]  42o5s7n                  ^zlib@1.2.11%cce@12.0.1+optimize+pic+shared patches=e93f7400712c2814905815204dadbdebcb91dc77dd586f60cbc82efa102fb539 arch=cray-sles15-zen2
[+]  o7jkiq5              ^ncurses@6.2%cce@12.0.1~symlinks+termlib abi=none arch=cray-sles15-zen2
[+]  ue5oxfq              ^tar@1.34%cce@12.0.1 arch=cray-sles15-zen2
[+]  nts3lqd      ^cray-mpich@8.1.7%cce@12.0.1 arch=cray-sles15-zen2
[+]  ipakrf7      ^elfutils@0.185%gcc@10.3.0~bzip2~debuginfod+nls~xz patches=d786d49c28d7f0c8fc27bab39ca8714e5f4d128c7f09bb18533a8ec99b38dbf8 arch=cray-sles15-zen2
[+]  kxlgpt5          ^m4@1.4.19%gcc@10.3.0+sigsegv patches=9dc5fbd0d5cb1037ab1e6d0ecc74a30df218d0a94bdd5a02759a97f62daca573,bfdffa7c2eb01021d5849b36972c069693654ad826c1a20b53534009a4ec7a89 arch=cray-sles15-zen2
[+]  y4jc7fv              ^libsigsegv@2.13%gcc@10.3.0 arch=cray-sles15-zen2
[+]  b3tnxxv      ^hwloc@2.5.0%cce@12.0.1~cairo~cuda~gl~libudev+libxml2~netloc~nvml~opencl+pci~rocm+shared arch=cray-sles15-zen2
[+]  r6njxzv          ^libpciaccess@0.16%cce@12.0.1 arch=cray-sles15-zen2
[+]  wuqmyi6              ^libtool@2.4.6%cce@12.0.1 arch=cray-sles15-zen2
[+]  d45x5xk              ^util-macros@1.19.3%cce@12.0.1 arch=cray-sles15-zen2
[+]  paafzoa      ^libdwarf@20180129%cce@12.0.1 arch=cray-sles15-zen2
[+]  3r3vmig      ^libunwind@1.5.0%cce@12.0.1~pic~xz~zlib arch=cray-sles15-zen2
[+]  aysjitm      ^otf2@2.2%cce@12.0.1 patches=6f22ce94964b4b9ceece610ec650a9ac5baea71317212f06c57136ce1e52e554,7e56d93b8ffe636cd7f298c7aa959a64ab5c8b3013e02320f71c2bbac5dd0653 arch=cray-sles15-zen2
[+]  ca2drud          ^autoconf@2.69%cce@12.0.1 patches=35c449281546376449766f92d49fc121ca50e330e60fefcfc9be2af3253082c2,7793209b33013dc0f81208718c68440c5aae80e7a1c4b8d336e382525af791a7,a49dd5bac3b62daa0ff688ab4d508d71dbd2f4f8d7e2a02321926346161bf3ee arch=cray-sles15-zen2
[+]  d7kgujc              ^perl@5.34.0%cce@12.0.1+cpanm+shared+threads arch=cray-sles15-zen2
[+]  flfq7jw                  ^berkeley-db@6.2.32%cce@12.0.1+cxx~docs+stl patches=b231fcc4d5cff05e5c3a4814f6a5af0e9a966428dc2176540d2c05aff41de522 arch=cray-sles15-zen2
[+]  hljobyl                  ^gdbm@1.19%cce@12.0.1 arch=cray-sles15-zen2
[+]  u77xev6                      ^readline@8.1%cce@12.0.1 arch=cray-sles15-zen2
[+]  4qrgyx7          ^automake@1.16.3%cce@12.0.1 arch=cray-sles15-zen2
[+]  o7tnjko      ^papi@6.0.0.1%cce@12.0.1~cuda+example~infiniband~lmsensors~nvml~powercap~rapl~sde+shared~static_tools patches=b6d6caa8c66a6495a24f32ba444e46f61e85083146dfa73ee3c8cf91fcfb4458 arch=cray-sles15-zen2
[+]  spjjgcs      ^pdt@3.25.1%cce@12.0.1~pic patches=113fca02f74cc49dc1add0f617a4bada116a999a1664598380e5e7652c8c231a arch=cray-sles15-zen2

```

**Comment by wspear** [comment]
The whole e4s enchilada: https://github.com/E4S-Project/e4s/blob/master/environments/21.08/spack.yaml
(Slight modifications to use the gcc version provided by Spock, avoid building the cuda specs and specify cray-mpich as the mpi provider)

**Comment by wspear** [comment]
I believe this is related to the mpich/cray-mpich configuration so I'm going to mention it here first. I'll create some package specific issues if the tagged maintainers prefer.

Several packages being built with spock's cray-mpich module as an external are failing with undefined reference errors like:

```
  >> 560    /usr/bin/ld: /opt/cray/pe/mpich/8.1.7/ofi/gnu/9.1/lib/libmpi_gnu_91.so: undefined reference to `fi_getinfo@@FABRIC_1.3'
  >> 561    /usr/bin/ld: /opt/cray/pe/mpich/8.1.7/ofi/gnu/9.1/lib/libmpi_gnu_91.so: undefined reference to `fi_version@@FABRIC_1.0'
  >> 562    /usr/bin/ld: /opt/cray/pe/mpich/8.1.7/ofi/gnu/9.1/lib/libmpi_gnu_91.so: undefined reference to `fi_freeinfo@@FABRIC_1.3'
  >> 563    /usr/bin/ld: /opt/cray/pe/mpich/8.1.7/ofi/gnu/9.1/lib/libmpi_gnu_91.so: undefined reference to `fi_strerror@@FABRIC_1.0'
  >> 564    /usr/bin/ld: /opt/cray/pe/mpich/8.1.7/ofi/gnu/9.1/lib/libmpi_gnu_91.so: undefined reference to `fi_fabric@@FABRIC_1.1'
  >> 565    /usr/bin/ld: /opt/cray/pe/mpich/8.1.7/ofi/gnu/9.1/lib/libmpi_gnu_91.so: undefined reference to `fi_dupinfo@@FABRIC_1.3'
```

I tried adding the fabric module to the mpich external but that didn't accomplish anything.

Affected packages from E4S are:
pcre: @tgamblin @adamjstewart (sorry, only 1 of these has official maintainers listed)
libzmq
gperftools: @albestro  @eschnett  @msimberg  @teonnik
xerces-c

**Comment by kwryankrattiger** [comment]
> Several packages being built with spock's cray-mpich module as an external are failing with undefined reference errors like:

I was running into this. I couldn't figure out why it happens, but the "fix" was to add the `libfabric` module  to all of the compilers. The other thing you may try is to add `libfabric` to the `cray-mpich` module list. The cray modules don't load all of the required modules with just a module load, so some need to be manually loaded, `libfabric` and `dmmsl` are two that I ran into, but there are probably more.

**Comment by christoph-conrads** [comment]
> Providing mpi with cray-mpich is preventing my environment from concretizing, presumably because one of the top level specs is forcing mpich. I haven't been able to track down which one yet.
> 
> ```
> ==> Error: The spec
>     'mpich ^findutils ^pkgconf'
>     is configured as not buildable, and no matching external installs were found
> ```

[`spack.yaml`, line 11](https://github.com/E4S-Project/e4s/blob/08808b78d51785ae9406b5d7e59bb520464480f4/environments/21.08/spack.yaml#L11) is at fault:
```yaml
  packages:
    all:
      compiler: [gcc@9.3.0]
      providers:
        blas: [openblas]
        mpi: [mpich]
```

**Comment by wspear** [comment]
> 
> 
> > Providing mpi with cray-mpich is preventing my environment from concretizing, presumably because one of the top level specs is forcing mpich. I haven't been able to track down which one yet.
> > ```
> > ==> Error: The spec
> >     'mpich ^findutils ^pkgconf'
> >     is configured as not buildable, and no matching external installs were found
> > ```
> 
> [`spack.yaml`, line 11](https://github.com/E4S-Project/e4s/blob/08808b78d51785ae9406b5d7e59bb520464480f4/environments/21.08/spack.yaml#L11) is at fault:
> 
> ```yaml
>   packages:
>     all:
>       compiler: [gcc@9.3.0]
>       providers:
>         blas: [openblas]
>         mpi: [mpich]
> ```

I am using a spack.yaml modified for spock:

```
spack:
  view: false
  concretization: separately
  packages:
    all:
      compiler:
      - gcc@10.3.0
      providers:
        blas:
        - openblas
        mpi:
        - cray-mpich
```

**Comment by wspear** [comment]
@kwryankrattiger that did the trick! I had already tried putting libfabric in the external mpi definition but putting it in the compiler has this working. Haven't run into any other missing modules yet.

**Comment by keitat** [comment]
Do these tricks still work on Spock?

**Comment by kwryankrattiger** [comment]
@keitat I have been using these still as of a week ago, so I assume yes unless something happened in the last week in spack/spock that would have changed this behavior.

**Comment by keitat** [comment]
I should talk to you offline. We use our own environment file.  I will try to introduce you environment setting. (https://github.com/xsdk-project/installxSDK/blob/r-0.7.0/platformFiles/spock/spack.yaml)

**Comment by tgamblin** [comment]
Do you still see this issue after #29392 ?

**Comment by becker33** [comment]
This issue should be fixed by a combination of 3 things:

1. #29392
2. The external spec should be `cray-mpich@...`, not `mpich@...`
3. Tau needs to be modified to not assume that every mpi implementation stores its headers in `$prefix/include`. That's not true for `cray-mpich`. Instead, Tau should use `spec["mpi"].headers.directories()` and similarly for the library directories.

**Comment by wspear** [comment]
@becker33 I wasn't aware of that design pattern and would like to switch, but it looks like headers.directories() returns an array. I'd guess that's usually just a single directory but if it is ever given more than one path TAU's configure script will break and doing headers.directories()[0] seems fragile. Can I assume directories[0] will always be where mpi.h/libmpi.so are located?

**Comment by becker33** [comment]
> Can I assume directories[0] will always be where mpi.h/libmpi.so are located?

I believe that is true of all current MPI providers, but you really can't assume that. Is this a limitation of the `tau` build system that it cannot handle an MPI implementation that requires multiple include directories?

**Comment by wspear** [comment]
Yes, though historically it hasn't really been a limitation. Tau can find what it needs as long as it has the location of mpi.h. Even now, multiple disjoint lib/include directories seems to be a theoretical problem.

**Comment by becker33** [comment]
You could always implement the search yourself, if you don't care what headers MPI thinks you need and only need `mpi.h`.

```
mpi_headers = llnl.util.filesystem.find_headers("mpi", self.spec["mpi"].prefix.include, recursive=True)
mpi_inc = mpi_headers.directories[0]
```
